q2_types_genomics/reference_db/_format.py

```diff
@@ -54,3 +60,167 @@ def _validate_(self, level):
+    def _validate_n_records(self, n=None):
+        with open(str(self), "r") as fh:
+            file_ = enumerate(fh) if n is None else zip(range(n), fh)
+
+            for i, line in file_:
+                line = line.rstrip("\n").split("\t|\t")
+                if len(line) != 4:
+                    raise ValidationError(
+                        "NCBI taxonomy names file must have 4 columns, "
+                        f"found {len(line)} columns on line {i + 1}."
+                    )
+                if not line[0].isnumeric():
+                    raise ValidationError(
+                        "NCBI taxonomy name file must contain a numeric "
+                        "taxonomy ID in the first column, found non-numeric "
+                        f"value on line {i + 1}: {line[0]}."
+                    )
+
+    def _validate_(self, level):
+        self._validate_n_records(n={"min": 10, "max": None}[level])
+
+
+class NCBITaxonomyBinaryFileFmt(model.BinaryFileFormat):
+    _accession_regex = re.compile(
+        r'[OPQ][0-9][A-Z0-9]{3}[0-9]|'  # UniProt
+        r'[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}|'  # UniProt
+        r'[A-Z]{3}\d{3,7}|'  # EMBL-EBI
+        r'[A-Z]+[-._]?\d+'  # NCBI
+    )
+    _accession_version_regex = re.compile(
+        r'[OPQ][0-9][A-Z0-9]{3}[0-9]\.\d+|'
+        r'[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}\.\d+|'
+        r'[A-Z]{3}\d{3,7}\.\d+|'
+        r'[A-Z]+[-._]?\d+\.\d+'
+    )
+    _taxid_regex = r'\d{1,10}'
+    _gi_regex = r'\d+'
+    _line_regex = re.compile(
+        rf"^({_accession_regex.pattern})\t"
+        rf"({_accession_version_regex.pattern})"
+        rf"\t({_taxid_regex})"
+        rf"\t({_gi_regex})\n$"
+    )
+
+    def _validate_1st_line(self, line: list):
+        if not (
+            line[0] == "accession" and
+            line[1] == "accession.version" and
+            line[2] == "taxid" and
+            line[3] == "gi"
+        ):
+            raise ValidationError(
+                "NCBI prot.accession2taxid file must have "
+                "columns: 'accession', 'accession.version' "
+                f", 'taxid' and 'gi'. Got {line} instead."
+            )
+
+    def _validate_Nth_line(self, line: list, line_no: int):
+        # For every filed validate one record
+        _line_regex = NCBITaxonomyBinaryFileFmt._line_regex
+        splitted_line = line.rstrip("\n").split(sep="\t")
+
+        # Raise exception if the entry does not match pattern
+        if not re.match(_line_regex, line):
+            raise ValidationError(
+                f"Non-allowed value found in line {line_no}.\n"
+                "Printing line:\n"
+                f"{splitted_line}"
+            )
+
+    def _validate_(self, level):
+        with gzip.open(str(self), 'rt', encoding='utf-8') as file:
+            # Flag first line
+            is_first_line = True
+            line_no = 1
+
+            # Set the maximum time for processing (in seconds)
+            max_processing_time = 60
```

Comment:
We should make use of the `level` parameter here and rather not focus on time here but maybe the specific count of lines. If the level is set to `max` we should validate a fixed count of lines (like, 1000? - I guess that can easily be validated within a minute? maybe more, if you have an estimate as to what is reasonable). If the level is `min` maybe check something like 10.

Reply:
In 1 minute it can parse ~16,304,548 lines. Should I set the max to 10,000,000?

Reply:
Sounds good!